Mathlib/RingTheory/LaurentSeries.lean

```diff
@@ -852,4 +876,300 @@ instance instLaurentSeriesComplete : CompleteSpace K⸨X⸩ :=
+    let X := {f : K⸨X⸩ | Valued.v f < ↑d}
+    refine ⟨(fun x : K⸨X⸩ × K⸨X⸩ ↦ x.snd - x.fst) ⁻¹' X, ⟨X, ?_⟩, ?_⟩
+    · refine ⟨?_, Set.Subset.refl _⟩
+      · simp only [Valued.mem_nhds, sub_zero]
+        use d
+    · refine subset_trans (fun _ _ ↦ ?_) pre_T
+      apply hd
+      erw [Set.mem_setOf_eq, sub_zero, valuation_eq_LaurentSeries_valuation,
+        coe_sub]
+      assumption
+
+theorem continuous_coe : Continuous (Coe.coe : RatFunc K → K⸨X⸩) :=
+  (isUniformInducing_iff'.1 (inducing_coe)).1.continuous
+
+/-- The `X`-adic completion as an abstract completion of `RatFunc K`-/
+abbrev ratfuncAdicComplPkg : AbstractCompletion (RatFunc K) :=
+  UniformSpace.Completion.cPkg
+
+variable (K)
+/-- Having established that the `K⸨X⸩` is complete and contains `RatFunc K` as a dense
+subspace, it gives rise to an abstract completion of `RatFunc K`.-/
+noncomputable def LaurentSeriesPkg : AbstractCompletion (RatFunc K) where
+  space := K⸨X⸩
+  coe := Coe.coe
+  uniformStruct := inferInstance
+  complete := inferInstance
+  separation := inferInstance
+  isUniformInducing := inducing_coe
+  dense := coe_range_dense
+
+instance : TopologicalSpace (LaurentSeriesPkg K).space :=
+  (LaurentSeriesPkg K).uniformStruct.toTopologicalSpace
+
+@[simp]
+theorem LaurentSeries_coe (x : RatFunc K) : (LaurentSeriesPkg K).coe x = (↑x : K⸨X⸩) :=
+  rfl
+
+/-- Reintrerpret the extension of `coe : RatFunc K → K⸨X⸩` as ring homomorphism -/
+abbrev extensionAsRingHom :=
+  UniformSpace.Completion.extensionHom (coeAlgHom K).toRingHom
+
+/-- An abbreviation for the `X`-adic completion of `RatFunc K` -/
+abbrev RatFuncAdicCompl := adicCompletion (RatFunc K) (Polynomial.idealX K)
+
+/- The two instances below make `comparePkg` and `comparePkg_eq_extension` slightly faster-/
+instance : UniformSpace (RatFuncAdicCompl K) := inferInstance
+instance : UniformSpace K⸨X⸩ := inferInstance
+
+/-- The uniform space isomorphism between two abstract completions of `ratfunc K` -/
+abbrev comparePkg : RatFuncAdicCompl K ≃ᵤ K⸨X⸩ :=
+  compareEquiv ratfuncAdicComplPkg (LaurentSeriesPkg K)
+
+lemma comparePkg_eq_extension (x : UniformSpace.Completion (RatFunc K)) :
+    (comparePkg K).toFun x = (extensionAsRingHom K (continuous_coe)).toFun x := rfl
+
+/-- The uniform space equivalence between two abstract completions of `ratfunc K` as a ring
+equivalence: this will be the *inverse* of the fundamental one.-/
+abbrev ratfuncAdicComplRingEquiv : RatFuncAdicCompl K ≃+* K⸨X⸩ :=
+  {comparePkg K with
+    map_mul' := by
+      intro x y
+      rw [comparePkg_eq_extension, (extensionAsRingHom K (continuous_coe)).map_mul']
+      rfl
+    map_add' := by
+      intro x y
+      rw [comparePkg_eq_extension, (extensionAsRingHom K (continuous_coe)).map_add']
+      rfl }
+
+/-- The uniform space equivalence between two abstract completions of `ratfunc K` as a ring
+equivalence: it goes from `K⸨X⸩` to `RatFuncAdicCompl K` -/
+abbrev LaurentSeriesRingEquiv : K⸨X⸩ ≃+* RatFuncAdicCompl K :=
```

Comment:
Can we upgrade this to an `AlgEquiv`?

Reply:
Not straight away, because the `AdicCompl` construction does not yield `K`-algebra on `RatFunc`, basically because it is a construction on rings and not on algebras. But it is an important point, and I have added a ToDo.